### DISCUSSION TITLE
Adding a `format` code plugin event

### DISCUSSION THREAD

**Discussion by overbit** [amplication/amplication#6795] [Ideas]
Currently `data-service-generator` contains logic to format with a consistent set of rules generated code files (here: https://github.com/amplication/amplication/blob/master/packages/data-service-generator/src/server/create-server.ts#L83).

This behaviour works fine but it only format these files following the default set of prettier rules.
When a user select a `formatter` plugin, all the generated code should be formatted using the user's defined format function (i.e. for prettier plugin the function will be the prettier `format` with the configuration saved in the plugin settings).

In order to do so, we will need to allow `formatter` plugins to define their specific `formatCode` function that will be executed by dsg as per today.

**Comment by yuval-hazaz** [comment]
That is a good idea, and it should also be very easy to implement
The format code function we use is from prettier 
![image](https://github.com/amplication/amplication/assets/43705455/cec96149-2432-4493-b811-de9914d2b21a)

and it accepts an Option object with all the relevant configuration

![image](https://github.com/amplication/amplication/assets/43705455/4a656bc8-7cb4-42ca-b875-88499138bccf)


We can expose an object on the context and allow any plugin to set its value on any event, so we dont even need a new event for that.

Imagine that - I work in an organization with a specific set of formatting rules and want to ensure the code generated by Amplication follows those rules. so, what i can do is:
1. Create a plugin called "My Organization Formatting". 
2. Handle any event I choose to change the formatting of the code on the context object (beforeCreateServer for example)

**Reply by overbit** [reply]
That would be easy enough, but what will happen to other formatters? We will lock formatting to only prettier and not supporting anything else

**Comment by mulygottlieb** [comment]
I like the concept of adding an event for "beforeFormatCode" and "afterFormatCode" that will allow us to apply any formatter (e.g. non prettier) with any settings. However, I am not clear how it will work with multiple formatters (multiple plugins that implement the formatting event) and even how it will work between a single formatter and the "defaultBehaviour" formatting - since every formatter will likely override the previous formatter. Should we limit it to a single formatter plugin and to assume it always means "skipDefaultBehaviour" ?
